### ISSUE TITLE
CPU stuck on dqlite actively freezing the machine

### ISSUE THREAD

**Issue by steve96961** [canonical/k8s-dqlite#213]
I have a microk8s Cluster running with 5 Control-Plane Nodes and 1 Worker Node. A Control-Plane Node t21 was suddenly NotReady and the server unreachable. Neither ICMP nor SSH or Zabbix Monitoring could reach it. Only a Hardware Reset would resolve this problem. 

Almost all Workloads on the Node were in an unknown State after this and would not schedule. Forcefully deleting all of them resolved this although workloads scheduling on this node would take their sweet time and even require another forceful delete of the pod (e.g. Ingress Controller).

Syslog:

```
Dec 13 05:20:18 t21 kernel: [3635042.040155] CPU: 11 PID: 3604314 Comm: k8s-dqlite Tainted: G        W    L    5.15.0-122-generic #132-Ubuntu
Dec 13 05:20:18 t21 kernel: [3635042.040158] Hardware name: To Be Filled By O.E.M. To Be Filled By O.E.M./B450 Pro4, BIOS H3.90F 05/19/2020
Dec 13 05:20:18 t21 kernel: [3635042.040160] RIP: 0010:xas_load+0x1d/0xd0
Dec 13 05:20:18 t21 kernel: [3635042.040165] Code: 95 c0 44 89 c0 e9 03 c7 94 00 0f 1f 00 55 48 89 e5 41 56 41 55 49 89 fd 41 54 53 48 83 ec 08 e8 89 f0 ff ff 48 89 c2 83 e2 03 <48> 83 fa 02 75 08 48 3d 00 10 00 00 77 11 48 83 c4 08 5b 41 5c 41
```
This log probably reoccurs over a 1000 times per second (or more).

Kernel Log:
```
Dec 13 05:07:18 t21 kernel: [3634262.037078] watchdog: BUG: soft lockup - CPU#3 stuck for 82s! [k8s-dqlite:3604314]
```

This goes on for a while, time always increasing.

CPU load before and after the Server was unreachable
![image](https://github.com/user-attachments/assets/bb5c7466-7932-4f5d-923b-9378d0d0d1b6)

**Comment by steve96961** [comment]
At the time of this incident i am running channel 1.30/stable (1.30.1) on Ubuntu 22.04 with an exception (I'll get to that in a minute).

I am still not sure why any of this happened but I believe to have at least found the cause for it, buckle up.

About two months ago (13.10.2024 to be exact) I decided to run a `do-release-upgrade` to get the nodes on this cluster to Ubuntu 24.04, starting with t20. I had tested this process on a few other servers on a different cluster running microk8s beforehand, no problems. After the upgrade was done and the kubelet still wasn't answering decided to check the server again. 
To my dismay I had to find out that microk8s wasn't only not running. It was nowhere to be found. In fact the entire snapd package had been removed along with all other snap data, so that's cool.

So I reinstall snapd, my snaps and try to get the node back in.

```
microk8s join 10.10.10.21:25000/6cff19a97b119bd055f825a3458de242/ee8c3a298b57 Contacting cluster at 10.10.10.21 Connection failed. Failed to retrieve dqlite cluster nodes: the joining node (10.10.10.20) is already known to dqlite (500).
```
Right, the node is still known in the cluster, so I run `microk8s remove-node t20` on another control-plane. Still can't join the cluster. I suspected the lack of `microk8s leave` on the affected node had something to do with it but I obviously couldn't execute that. I didn't know the best way to go about this at the time, perhaps I should have changed and joined with a different hostname or tried restoring the dqlite data on this  node. I decided to interact directly with dqlite and remove the remaining t20  node.

Attempted to join again  and it works. Everything seems to be in order and there were no problems for almost two months. I obviously didn't proceed with the Ubuntu Upgrade on the remaining nodes after what happened.

Fast forward to two weeks ago (07.12.2024) t20 is suddenly unreachable. I didn't take any action and the Server was up and running again by itself 17h(!!!) later (except I had to restart microk8s services). What's odd to me is that during this 17h downtime the server didn't log  anything, no syslog, no kernel.log, just nothingness.

Then almost a week later it was this server (t21). This was a more pressing issue so i took immediate action (as described above). As this server had syslogs all the way I didn't correlate these two at first.

A few days ago I noticed some inconsistencies in my cluster when communicating with the kubernetes service.
e.g. when spamming `kubectl get no` inside a pod in the cluster I would mostly get a response instantly but sometimes I would't and instead timeout after a while. If I canceled this longer request and immediately fired again I'd get a response. This caused some major issues in my cluster with some helm upgrade hooks timing out because they were creating pods and executing `kubectl` commands inside that pod (Gitlab does this).

After removing the t20 node from the cluster pod network instability went away immediately and I have a feeling this was all connected.

Hope this helps.

[inspection-report-20241213_105536.tar.gz](https://github.com/user-attachments/files/18194424/inspection-report-20241213_105536.tar.gz)

**Comment by louiseschmidtgen** [comment]
Hi @steve96961,
Thank you for the additional context.

When upgrading the underlying OS of a node, the safest way is to add a new node with the new OS version, drain and cordon the node with the old OS version and then remove it from the cluster.

In your case, if your environment allows it, I would recommend adding a fresh node as described above and removing the `t20` node all-together.

Would that be an option for you?

**Comment by steve96961** [comment]
Yes that is the way I have done it. It would be nice if there was a "cleaner" way to upgrade a node but I will be doing it this way for now.

 Still curious about what exactly caused the dqlite to fry my cpu but after removing the node there have been no more issues as of now.

_Guten Rutsch_

**Comment by louiseschmidtgen** [comment]
Hello @steve96961,

I'm glad to hear that your cluster is stable now. Doing a node upgrade can potentially lead to issues, therefore the safe way (drain and cordon nodes) is also the recommended procedure by upstream k8s.

_Frohes Neues!_
